### PR TITLE
abci/client: simplify client interface

### PR DESCRIPTION
**Pull request by creachadair**
As we discussed at the ABCI++ meeting, the rollout in v0.36 is an ideal time to clean up some  infelicities in the original implementation. This is a first step toward that cleanup. We will still have to address Flush, CommitTx, and DeliverTx separately, but getting the easy cases out of the way will help keep those diffs smaller.

N.B.: The diff in `grpc_client.go` is large, but less complicated than it looks. It folds the sync/async versions back together, strips out the unnecessary support code, and updates the names. The rename makes the diff look hairier than it should be.

This change has two main effects:

1. Remove most of the Async methods from the abci.Client interface.
   Remaining are FlushAsync, CommitTxAsync, and DeliverTxAsync.

2. Rename the synchronous methods to remove the "Sync" suffix.

The rest of the change is updating the implementations, subsets, and mocks of
the interface, along with the call sites that point to them.

TODO before merge:
- [ ] Update changelog.

### CODE REVIEW

**Review by marbar3778 — COMMENTED** [review]
love to see this being cleanup. Before merging might want to do a sanity check with @ebuchman, I believe he may know why this exists. 

Im for this change and in the sdk we only use the local client, so not sure where all this came in

File: abci/client/client.go
Comment (cmwaters) [inline abci/client/client.go:35]: Why do we still need this?

File: abci/client/client.go
Comment (cmwaters) [inline abci/client/client.go:37]: I believe this gets used in the mempool right? Is it possible (in a follow up PR of course) to modify the usage so we can call `CheckTx` instead?

**Review by cmwaters — APPROVED** [review]
LGTM pending a check from @ebuchman or @milosevic

File: abci/client/client.go
Comment (tychoish) [reply abci/client/client.go:37]: Yeah, I tried this, sort of naively, and I think it's possible, but the impact on the mempool is pretty good (mostly because there are a bunch of tests that do assertions/etc. via the callback. I think the best path forward would be to merge this, and then write a shim in the mempool package that does the switch and try and go from there.

File: abci/client/client.go
Comment (creachadair) [reply abci/client/client.go:35]: We might not. It's called in the [mempool](https://github.com/tendermint/tendermint/blob/mjf/desync/internal/mempool/mempool.go#L738) and the [proxy app](https://github.com/tendermint/tendermint/blob/mjf/desync/internal/proxy/app_conn.go#L142). It might be we could get rid of those, or swap them to synchronous, but I didn't want to expand this change with functional changes.

Comment (cmwaters) [reply abci/client/client.go:35]: Yeah makes sense to tackle this as a different issue/PR

**Review by tychoish — APPROVED** [review]

**Comment by cmwaters** [comment]
FYI, I followed up with Ethan on slack and he had this response:
> this has never been benchmarked but jaes original intention was performance, thinking especially about out-of-process apps. the idea being that txs could quickly sent to the app without having to wait for response
>i think its only used for delivertx and checktx. the whole mempool was built around it. does it even work for in-process apps i dont rememver? i think so
>so conceivably could be a performance hit in the mempool to use sync abci instead but its also possible the server side abci and sdk arent even making use of it properly and so its all moot lol

**Review by williambanfield — COMMENTED** [review]
This makes me very happy. Not sure I understand the distinction between Flush vs FlushAsync. Didn't look at the mocks.

File: abci/client/socket_client.go
Comment (creachadair) [reply abci/client/socket_client.go]: In re: Flush vs. FlushAsync. As far as I can tell, there is no difference, but when I tried to get rid of FlushAsync it metastasized. I suspect there is a kernel of simplicity inside that we could use. (Flush appears to be primarily an artifact of the socket client anyway, cf. #6994).

In re: Plumbing methods. Sure, done. But it's harder to be systematic about those.

```
gofmt -r 'queueRequestAndFlushSync->queueRequestAndFlush' -w  abci/client/socket_client.go
```

**Comment by creachadair** [comment]
> FYI, I followed up with Ethan on slack and he had this response:
> 
> > this has never been benchmarked but jaes original intention was performance, thinking especially about out-of-process apps. the idea being that txs could quickly sent to the app without having to wait for response
> > i think its only used for delivertx and checktx. the whole mempool was built around it. does it even work for in-process apps i dont rememver? i think so
> > so conceivably could be a performance hit in the mempool to use sync abci instead but its also possible the server side abci and sdk arent even making use of it properly and so its all moot lol

I agree this is something worth measuring. At the moment I think neither in-process nor out-of-process applications can practically benefit from concurrency here: Even for CheckTx, the socket protocol requires head-of-line blocking because requests cannot be interleaved. And for in-process all client requests share a global mutex. Both of those are issues I think we should address—for performance, certainly, but also to make the control flow easier to understand.

For now, though, I don't think these client-side changes can have any material effect on the performance (in either direction).

**Review by williambanfield — APPROVED** [review]

**Pull request merged**
